### PR TITLE
Provide sensible defaults for GuLambda memorySize and timeout

### PR DESCRIPTION
**Pull request by jacobwinch**
## What does this change?

Whilst testing https://github.com/guardian/cdk/pull/186, I noticed that `memorySize` and `timeout` are optional parameters for a `GuLambda`. If these values are omitted, CloudFormation defaults to [128MB](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-memorysize) and [3 seconds](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-timeout), respectively.

I don't think that these default values are particularly helpful for many of the Guardian use-cases that I've seen (they are especially unhelpful for Scala lambdas!), so I've tried to suggest something more appropriate.

## Does this change require changes to existing projects or CDK CLI?

There are no required changes, but some projects may automatically increase the memory or timeout values for lambda functions. This change would show up in a snapshot diff though, so people could set these low values again manually if they felt strongly about keeping them.

## How to test

I think this is adequately covered by unit tests added in this PR.

## How can we measure success?

When setting projects up from scratch developers should save time, as they are less likely to encounter timeouts or OOM errors immediately after deploying for the first time.

## Have we considered potential risks?

I suppose there is a small risk that this could incur unnecessary costs (due to the way that [lambda pricing](https://aws.amazon.com/lambda/pricing/) works).

There is also a risk that we will over or under provision lambdas due to my assumptions about what 'sensible' defaults are. An alternative approach would be to make these properties mandatory for a `GuLambda`. This would force developers to make these choices up-front, but at least it would avoid the annoying case where you end up with the super-low CloudFormation defaults!

### CODE REVIEW

File: src/constructs/lambda/lambda.ts [hunk 1]
```diff
@@ -21,12 +22,28 @@ interface GuFunctionProps extends Omit<FunctionProps, "code"> {
   apis?: ApiProps[];
 }
 
+function defaultMemorySize(runtime: Runtime, memorySize?: number): number {
+  if (memorySize) {
+    return memorySize;
+  } else {
+    switch (runtime.family) {
+      case RuntimeFamily.JAVA:
+        return 1024;
```
Comment (jacobwinch) [inline src/constructs/lambda/lambda.ts:31]: I based this on https://github.com/guardian/lambda-scala-boilerplate.g8/blob/master/src/main/g8/cfn.yaml#L67.

File: src/constructs/lambda/lambda.ts [hunk 2]
```diff
@@ -21,12 +22,28 @@ interface GuFunctionProps extends Omit<FunctionProps, "code"> {
...
+      default:
+        return 512;
```
Comment (jacobwinch) [inline src/constructs/lambda/lambda.ts:33]: I think the other main family that we want to cater for is `NODE`, so I've based this on https://github.com/guardian/tag-janitor/blob/main/cdk/lib/cdk-stack.ts#L54 and https://github.com/guardian/google-chat-bots/blob/main/cdk/lib/constants.ts#L29.

File: src/constructs/lambda/lambda.ts [hunk 3]
```diff
@@ -21,12 +22,28 @@ interface GuFunctionProps extends Omit<FunctionProps, "code"> {
...
+    }
+  }
+}
+
 export class GuLambdaFunction extends Function {
   constructor(scope: GuStack, id: string, props: GuFunctionProps) {
     const bucket = Bucket.fromBucketName(scope, `${id}-bucket`, props.code.bucket);
     const code = Code.fromBucket(bucket, props.code.key);
+    const { memorySize, timeout, ...rest } = props;
```
Comment (jamie-lynch) [inline src/constructs/lambda/lambda.ts:42]: Do these still need to be extracted from props or would it be easier just to use `props.memorySize` on line 45?

File: src/constructs/lambda/lambda.ts [hunk 4]
```diff
@@ -21,12 +22,28 @@ interface GuFunctionProps extends Omit<FunctionProps, "code"> {
...
     super(scope, id, {
-      ...props,
+      ...rest,
+      memorySize: defaultMemorySize(props.runtime, memorySize),
+      timeout: props.timeout ? props.timeout : Duration.seconds(30),
```
Comment (jamie-lynch) [inline src/constructs/lambda/lambda.ts:46]: Minor but the following should achieve the same result with marginally less code
```suggestion
      timeout: props.timeout ?? Duration.seconds(30),
```

File: src/constructs/lambda/lambda.ts [hunk 5]
```diff
@@ -21,12 +22,28 @@ interface GuFunctionProps extends Omit<FunctionProps, "code"> {
...
+function defaultMemorySize(runtime: Runtime, memorySize?: number): number {
+  if (memorySize) {
+    return memorySize;
+  } else {
+    switch (runtime.family) {
+      case RuntimeFamily.JAVA:
+        return 1024;
+      default:
+        return 512;
+    }
```
Comment (jamie-lynch) [inline src/constructs/lambda/lambda.ts:34]: Nice 👍

File: src/constructs/lambda/lambda.ts (hunk 3 above)
Comment (jacobwinch) [reply src/constructs/lambda/lambda.ts]: Ah yes thanks, I didn't realise this simpler version would work (but now I've read the docs and re-tested I can see that my version was needlessly complicated 😀).

**Review by jamie-lynch — APPROVED** [review]

**Pull request merged**